storage-proofs-update/src/vanilla.rs

```diff
@@ -0,0 +1,1183 @@
+                [apex_leafs_height..apex_leafs_height + apex_select_bit_len]
+                .iter()
+                .zip(apex_tree.iter())
+                .all(|(path_elem, apex_tree_row)| apex_tree_row.contains(&path_elem.0[0]))
+            {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn verify_all_partitions(
+        pub_params: &Self::PublicParams,
+        pub_inputs: &Self::PublicInputs,
+        partition_proofs: &[Self::Proof],
+    ) -> Result<bool> {
+        if partition_proofs.len() != pub_params.partition_count {
+            return Ok(false);
+        }
+        for (k, partition_proof) in partition_proofs.iter().enumerate() {
+            let mut partition_pub_inputs = pub_inputs.clone();
+            partition_pub_inputs.k = k;
+            if !Self::verify(pub_params, &partition_pub_inputs, partition_proof)? {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    fn with_partition(mut pub_inputs: Self::PublicInputs, k: Option<usize>) -> Self::PublicInputs {
+        pub_inputs.k = k.unwrap();
+        pub_inputs
+    }
+}
+
+// `phi = H(comm_d_new, comm_r_old)`
+pub fn phi(comm_d_new: &TreeDDomain, comm_r_old: &TreeRDomain) -> TreeRDomain {
+    let comm_d_new: TreeRDomain = {
+        let comm_d_new: Fr = (*comm_d_new).into();
+        comm_d_new.into()
+    };
+    <TreeRHasher as Hasher>::Function::hash2(&comm_d_new, &comm_r_old)
+}
+
+fn mmap_read(path: &Path) -> Result<Mmap, Error> {
+    let f_data = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .with_context(|| format!("could not open path={:?}", path))?;
+    unsafe {
+        MmapOptions::new()
+            .map(&f_data)
+            .with_context(|| format!("could not mmap path={:?}", path))
+    }
+}
+
+fn mmap_write(path: &Path) -> Result<MmapMut, Error> {
+    let f_data = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .open(path)
+        .with_context(|| format!("could not open path={:?}", &path))?;
+    unsafe {
+        MmapOptions::new()
+            .map_mut(&f_data)
+            .with_context(|| format!("could not mmap path={:?}", path))
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+#[allow(clippy::from_iter_instead_of_collect)]
+impl<TreeR> EmptySectorUpdate<TreeR>
+where
+    TreeR: 'static + MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    /// Returns tuple of (comm_r_new, comm_r_last_new, comm_d_new)
+    pub fn encode_into(
+        nodes_count: usize,
+        tree_d_new_config: StoreConfig,
+        tree_r_last_new_config: StoreConfig,
+        comm_c: TreeRDomain,
+        comm_r_last_old: TreeRDomain,
+        new_replica_path: &Path,
+        new_cache_path: &Path,
+        sector_key_path: &Path,
+        sector_key_cache_path: &Path,
+        staged_data_path: &Path,
+        h: usize,
+    ) -> Result<(TreeRDomain, TreeRDomain, TreeDDomain)> {
+        // Sanity check all input path types.
```

Comment:
May be also ensure that the `new_replica_path` (and cache path) is not equal to the old one here?